libvast/vast/system/source.hpp

```diff
@@ -257,46 +263,35 @@ source(caf::stateful_actor<source_state<Reader>>* self, Reader reader,
                                             push_slice);
       t.stop(produced);
       st.count += produced;
-      auto force_emit_batches = [&] {
-        st.mgr->out().fan_out_flush();
-        st.mgr->out().force_emit_batches();
-        st.send_report();
-      };
       auto finish = [&] {
         st.done = true;
         st.send_report();
         self->quit();
       };
-      if (produced == 0) {
-        // If the source is unable to generate new events (returns 0), the
-        // source will stall and never be polled again, because CAF expects
-        // stream sources to _always_ be able to generate new events. As a
-        // workaround, we send out an invalid table slice here that gets ignored
-        // downstream, and forcefully emit batches.
-        VAST_VERBOSE(self, "emits invalid table slice to prevent stalling");
-        out.push(nullptr);
-        return finish();
-      }
       if (st.requested && st.count >= *st.requested)
         return finish();
-      if (err != caf::none) {
-        if (err == vast::ec::timeout) {
-          if (produced > 0) {
-            VAST_DEBUG(self, "hit input timeout and forcefully emits", produced,
-                       "produced events");
-            return force_emit_batches();
-          } else {
-            // This case should never happen. If it does, we hit an internal
-            // application logic error in the reader. CAF might stall out on us
-            // here.
-            VAST_ERROR(self, "hit input timeout, but produced no events");
-            return finish();
-          }
-        } else {
-          if (err != vast::ec::end_of_input)
-            VAST_INFO(self, "completed with message:", render(err));
-          return finish();
+      if (err == ec::timeout) {
+        if (!st.waiting_for_input) {
+          // This pull handler was invoked while we were waiting for a wakeup
+          // message. Sending another one would create a parallel wakeup cycle.
+          st.waiting_for_input = true;
+          self->delayed_send(self, st.wakeup_delay, atom::wakeup_v);
+          // Exponential backoff for the wakeup calls.
+          // For each consecutive invocation of this generate handler that does
+          // not emit any events, we double the wakeup delay.
+          // The sequence is 0, 20, 40, 80, 160, 320, 640, 1280.
+          if (st.wakeup_delay == std::chrono::milliseconds::zero())
+            st.wakeup_delay = std::chrono::milliseconds{20};
```

Comment:
Is it just coincidence that we're using the same initial value here as the read timeout, or should this be configurable via `import.read-timeout` as well?

I personally think it is fine to leave these fixed.

Reply:
It is coincidence. But we should cap the maximum at `batch-timeout / 2`.

Reply:
That is a sensible choice.

Reply:
This is done now.